src/main/java/org/apache/datasketches/tuple/AnotB.java

```diff
@@ -22,45 +22,327 @@
+    final CompactSketch<S> cskA = (skA instanceof CompactSketch)
+        ? (CompactSketch<S>)skA
+        : ((QuickSelectSketch<S>)skA).compact();
+    keys_ = cskA.keys_;
+    summaries_ = cskA.summaries_;
+    count_ = cskA.getRetainedEntries();
+  }
+
+  /**
+   * Performs an <i>AND NOT</i> operation with the existing internal state of this AnoB operator.
+   * @param skB The incoming Tuple sketch for the second (or following) argument <i>B</i>.
+   */
+  @SuppressWarnings("unchecked")
+  public void notB(final Sketch<S> skB) {
+    if (empty_) { return; }
+    if ((skB == null) || skB.isEmpty()) { return; }
+    //skB is not empty
+    final long thetaLongB = skB.getThetaLong();
+    thetaLong_ = Math.min(thetaLong_, thetaLongB);
+    //Build hashtable and removes keys of skB >= theta
+    final int countB = skB.getRetainedEntries();
+    final long[] hashTableKeysB = convertToHashTable(skB.keys_, countB, thetaLong_);
+
+    //build temporary arrays of skA
+    final long[] tmpKeysA = new long[count_];
+    final S[] tmpSummariesA =
+        (S[]) Array.newInstance(summaries_.getClass().getComponentType(), count_);
+
+    //search for non matches and build temp arrays
+    int nonMatches = 0;
+    for (int i = 0; i < count_; i++) {
+      final long key = keys_[i];
+      if ((key != 0) && (key < thetaLong_)) { //skips keys of A >= theta
+        final int index =
+            HashOperations.hashSearch(hashTableKeysB, simpleIntLog2(hashTableKeysB.length), key);
+        if (index == -1) {
+          tmpKeysA[nonMatches] = key;
+          tmpSummariesA[nonMatches] = summaries_[i];
+          nonMatches++;
+        }
+      }
+    }
+    keys_ = Arrays.copyOfRange(tmpKeysA, 0, nonMatches);
+    summaries_ = Arrays.copyOfRange(tmpSummariesA, 0, nonMatches);
+    count_ = nonMatches;
+  }
+
+  /**
+   * Performs an <i>AND NOT</i> operation with the existing internal state of this AnoB operator.
+   * @param skB The incoming Theta sketch for the second (or following) argument <i>B</i>.
+   */
+  @SuppressWarnings("unchecked")
+  public void notB(final org.apache.datasketches.theta.Sketch skB) {
+    if (empty_) { return; }
+    if ((skB == null) || skB.isEmpty()) { return; }
+    //skB is not empty
+    final long thetaLongB = skB.getThetaLong();
+    thetaLong_ = Math.min(thetaLong_, thetaLongB);
+    //Build hashtable and removes keys of skB >= theta
+    final int countB = skB.getRetainedEntries();
+    final long[] hashTableKeysB =
+        convertToHashTable(extractThetaHashArray(skB, countB), countB, thetaLong_);
+
+    //build temporary arrays of skA
+    final long[] tmpKeysA = new long[count_];
+    final S[] tmpSummariesA =
+        (S[]) Array.newInstance(summaries_.getClass().getComponentType(), count_);
+
+    //search for non matches and build temp arrays
+    int nonMatches = 0;
+    for (int i = 0; i < count_; i++) {
+      final long key = keys_[i];
+      if ((key != 0) && (key < thetaLong_)) { //skips keys of A >= theta
```

Comment:
Thank you for these comments and bring it on!  

Hashes are 64 bits. The likelihood that a zero hash would occur is extremely unlikely. Assuming the hash function is perfect in its uniform random properties (which it is not!), a zero hash would occur about once in about 10^19 events.  There is no way to create a hash table of hashes where the empty slot = 0, if zero hashes were significant.  Currently, if a zero hash actually did occur it is effectively ignored, since inserting a zero into an empty slot results in the slot still being empty.   

So how significant is that?  Remember that these are probabilistic algorithms where the RSE on even the largest sketch (with the smallest error) is many orders of magnitude larger than the error that would be produced by ignoring hashes that happen to be zero because they are so rare.

Reply:
This calls into question whether the check (key != 0) is even necessary.  In theory, probably not.  By the way, hash keys in the theta family are forced to be positive or zero by shifting the output of the hash function right by one bit (effectively making them 63 bits of entropy).  Actually this check, if it exists at all, should probably be (key > 0).

So why do I put in checks like this?  Because I am conservative and worried about the other source of non-valid hashes, which is corruption of the sketch binaries as they are moved around in a system. Which, by the way, is far more likely than a valid hash being <= 0.  So this little check just insures that invalid hashes don't get further propagated.

Reply:
I hadn't considered the case where the contents of the sketch itself are subject to corruption.  I agree that it is valid to preserve the check in this case. 
It's very useful background information, and it's interesting to see the actual probability explained in those terms and contrasted to the actual RSE of the sketch.  Thanks for the explanation!